### ISSUE TITLE
Add a `Serialize::serialize_into` function

### ISSUE THREAD

**Issue by eduardosm** [psychon/x11rb#213]
Currently, serializing structures with variable-length nested structures with first serialized the nested structure into a `Vec`, concatenate it into the parent structure `Vec` and discard the nested structure `Vec`:

For example, in `xinput::DeviceClass::serialize`:
```rust
result.extend(self.data.serialize());
```

`self.data.serialize()` will produce a `Vec`, and `result.extend()` will append it to `result` and discard it.

To reduce the number of allocations and byte copies, I propose to add an additional function to the `Serialize` trait:

```rust
fn serialize_into(&self, bytes: &mut Vec<u8>);
```

This function, instead of producing a `Vec<u8>` (or a `Self::Bytes`), will directly append the serialized structure into `bytes`.

If you agree with it, I can implement it.

**Comment by psychon** [comment]
Does this just "feel inefficient" to you or did you come across this as a problem? I am not sure how much of a problem this actually is. Yes, this can cause extra allocations, but in a network protocol like that, I am not sure how much of a problem this actually is. Still, if you want to do the work, I am not going to stop you (and I also optimised some stuff already, e.g. `Serialize::Bytes` is a pure optimisation and could be replaced with `Vec<u8>`).

What do you propose as the new form of the `Serialize` trait? Do you want to replace the `serialize` method with `serialize_into`? Or should both functions exist? should one of the functions be provided as a default function based on the other? Without generic associated types, I am not sure how to express what I have in mind for the trait...

**Comment by eduardosm** [comment]
It just "feels inefficient".

The form I propose is:
```rust
pub trait Serialize {
    type Bytes;
    fn serialize(&self) -> Self::Bytes;
    fn serialize_into(&self, bytes: &mut Vec<u8>);
}
```

Depending on each situation, `serialize` would be implemented on top of `serialize_into` or `serialize_into` on top of `serialize`.

For example, for fixed length structures:

```rust
impl Serialize for Whatever {
    type Bytes = [u8; N];
    fn serialize(&self) -> Self::Bytes { ... }
    fn serialize_into(&self, bytes: &mut Vec<u8>) {
        bytes.extend_from_slice(&self.serialize());
    }
}
``` 

And for variable length structures:

```rust
impl Serialize for Whatever {
    type Bytes = Vec<u8>;
    fn serialize(&self) -> Self::Bytes {
        let mut result = Vec::new();
        self.serialize_into(&mut result);
        result
    }
    fn serialize_into(&self, bytes: &mut Vec<u8>) { ... }
}
```

The `Serialize` for `[T]` (https://github.com/psychon/x11rb/blob/1e6289fb7ef20041038b8fca440520bd731aedac/src/x11_utils.rs#L276-L288) would become:

```rust
impl<T: Serialize> Serialize for [T] {
    type Bytes = Vec<u8>;
    fn serialize(&self) -> Self::Bytes {
        let mut result = Vec::new();
        self.serialize_into(&mut result);
        result
    }
    fn serialize_into(&self, bytes: &mut Vec<u8>) {
        for item in self {
            item.serialize_into(bytes);
        }
    }
}
```
In that case, the `AsRef<[u8]>` bound would not be needed.

**Comment by psychon** [comment]
Okay, no `default` functions in the traits. Since this is basically only implemented by the code generator that makes sense. So: :+1: 

Thank you for the suggestions.

**Comment by psychon** [comment]
Fixed by #216.
